File: Plugin/MessageClient.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.IO;
using System.IO.Pipes;
using System.Text;

namespace HandleAbuserPlugin
{
	internal class MessageClient : IDisposable
	{
		private readonly PipeStream pipe;

		private readonly Dictionary<MessageType, Func<IMessage>> registeredMessages = new Dictionary<MessageType, Func<IMessage>>();

		public MessageClient(PipeStream pipe)
		{
			Contract.Requires(pipe != null);

			this.pipe = pipe;
		}

		public void Dispose()
		{
			pipe?.Dispose();
		}

		public void RegisterMessage<T>() where T : IMessage, new()
		{
			IMessage MessageCreator() => new T();

			registeredMessages.Add(MessageCreator().MessageType, MessageCreator);
		}

		public IMessage Receive()
		{
			using (var ms = new MemoryStream())
			{
				var buffer = new byte[256];
				do
				{
					var length = pipe.Read(buffer, 0, buffer.Length);
					ms.Write(buffer, 0, length);
				}
				while (!pipe.IsMessageComplete);

				ms.Position = 0;

				using (var br = new BinaryReader(ms, Encoding.Unicode, true))
				{
					var type = (MessageType)br.ReadInt32();

					if (registeredMessages.TryGetValue(type, out var createFn))
					{
						var message = createFn();
						message.ReadFrom(br);
						return message;
					}
				}
			}

			return null;
		}

		public void Send(IMessage message)
		{
			Contract.Requires(message != null);

			using (var ms = new MemoryStream())
			{
				using (var bw = new BinaryWriter(ms, Encoding.Unicode, true))
				{
					bw.Write((int)message.MessageType);
					message.WriteTo(bw);
				}

				var buffer = ms.ToArray();
				pipe.Write(buffer, 0, buffer.Length);
			}
		}
	}
}

File: PipeServer/MessageClient.cpp
#include "MessageClient.hpp"
#include "PipeStream/MemoryStream.hpp"

MessageClient::MessageClient(PipeStream& _pipe)
	: pipe(_pipe)
{

}
//---------------------------------------------------------------------------
std::unique_ptr<IMessage> MessageClient::Receive()
{
	MemoryStream ms;
	std::vector<uint8_t> buffer(256);
	do
	{
		const auto length = pipe.Read(buffer.data(), 0, static_cast<int>(buffer.size()));
		ms.Write(buffer.data(), 0, length);
	} while (!pipe.IsMessageComplete());

	ms.SetPosition(0);

	BinaryReader br(ms);
	const auto type = static_cast<MessageType>(br.ReadInt32());

	const auto it = registeredMessages.find(type);
	if (it != std::end(registeredMessages))
	{
		auto message = it->second();
		message->ReadFrom(br);

		return message;
	}

	return nullptr;
}
//---------------------------------------------------------------------------
void MessageClient::Send(const IMessage& message) const
{
	MemoryStream ms;
	BinaryWriter bw(ms);

	bw.Write(static_cast<int>(message.GetMessageType()));
	message.WriteTo(bw);

	auto data = ms.ToArray();
	pipe.Write(data.data(), 0, static_cast<int>(data.size()));
}
//---------------------------------------------------------------------------
